text/vga-console.md

```diff
@@ -0,0 +1,106 @@
+apparently-atomic updates to the screen, although this operation need not be
+*actually* atomic as (a) good design and Rust checks should guarantee only one
+VGA structure owns video memory, and (b) moving 4k of memory is a long time to
+not have any other interrupts.
+
+### Proposed Design
+
+Philipp Oppermann suggests using a Unique pointer to wrap the magic memory
+location so that the compiler can guarantee that there is ever only one mutable
+owner of the video memory. This is a good plan, but in practice may be overkill
+for intermezzOS' needs and goals. Unique is yet another feature of Rust nightly
+that needs to be set as a crate-level attribute, and while intermezzOS already
+makes plentiful use of such features, I feel that where some can be trimmed, we
+may want to do so.
+
+I propose designing a VGA structure which owns only a single VGA-memory-sized
+slice, rather than the current implementation which gives one VGA structure two
+regions of memory. Since the magic address is the kind of implementation detail
+which should be hidden from consumers, the VGA structure will implement multiple
+constructors: the standard `new()` method will take a generic buffer, which can
+be allocated anywhere in normal memory (but as intermezzOS does not yet have a
+dynamic allocator, it will probably be in BSS), and an `at_vga_memory()` method
+will create a VGA structure that has taken ownership of the video memory region.
+
+In order to achive the pseudo-atomic functionality currently present in our
+console implementation, the VGA structure will need to implement a method that
+will permit one structure to borrow another and copy the borrowed buffer into
+its own storage. When the VGA structure at video memory does this, we achieve
+pseudo-atomic video update, and we will also be able to rapidly change display
+contexts by switching between multiple different VGA structures in main memory.
+
+### Other Considerations
+
+From the intermezzOS workshop and from personal experience, I noticed that
+forgetting some details of the VGA buffer's size was a common source of bugs.
+Mixing up rows vs columns, or forgetting that each cell was two bytes wide not
+one, leads to some interesting learning opportunities in debugging.
+
+Some of the problems with C that Rust aims to overcome are type safety and
+buffer overruns. Working with video memory provides opporunities to showcase
+Rust's power in both regards.
+
+While we obviously can continue to treat VGA memory as a single chunk of 4000
+`u8`s, I feel that doing so offers no benefit to writing intermezzOS in C. Mr.
+Oppermann's tutorial defines a character struct two bytes wide that internally
+handles ordering (little- vs big- endianness is a frequent bug) and sizing, so
+the screen can be represented in terms of these two-byte character structs
+rather than individual bytes, eliminating the need to remember to multiply the
+memory size by two.
```

Comment:
This is definitely something that we want to do, but I wasn't sure how much Rust experience people had in the workshop, and how much time... but big :+1: